### ISSUE TITLE
Sometimes lint discarded_features does not recognize `unwaited`.

### ISSUE THREAD

**Issue by polina-c** [dart-lang/linter#3801]
PR: https://github.com/flutter/devtools/pull/4659

**Comment by lrhn** [comment]
The lint, as described by itself, does not make exceptions for `unawaited`. It asks you to make the surrounding function `async` because you are calling an `async` function. If you do that, then you can use `unawaited` to avoid the `unawaited_futures` lint, which only applies to `async` functions.

Maybe `discarded_futures` lint should make an exception for futures which are not discarded. That is, it could use the same rules as `unawaited_futures`, only in a non-async function, and only complain if the future is actually discarded. If the future is used in a way that does not create a new future, then forcing the surrounding function to be async won't actually do anything useful. It will just force callers of that function to do the same thing.


(It's not a lint I'd ever consider enabling myself, because there are plenty of reasons to make non-`async` functions calling `async` functions when you are writing library code which works with futures. 
For application code, it's more likely that you will only be consuming futures created by other code.)

**Comment by oprypin** [comment]
Disregarding high-level considerations, I think there can't be any argument made that this lint is **better** by still complaining even after one adds `unawaited`. This is a false positive and should be fixed.

**Comment by oprypin** [comment]
Actually never mind, I'm not sure what actually happens with the interaction with `RecursiveAstVisitor`

**Comment by lrhn** [comment]
Yup, the `unawaited` function returns `void`, so that's not an issue. (https://api.dart.dev/stable/2.14.1/dart-async/unawaited.html).

There is nothing magical about the `unawaited` function, the `unawaited_futures` lint doesn't need to recognize it, because that lint just looks for futures which are not processed in some way. Calling `unawaited` function counts as processing the argument, and since it returns `void`, there is nothing more to check for.

That's another reason this lint worries me, it has to recognize `unawaited` explicitly, and therefore won't work with other, equally valid, ways to process a future (like `Future.ignore()`).

**Comment by pq** [comment]
@oprypin: I agree 100% and would love to see `@awaitNotRequired` or similar implemented.  Maybe time to revisit https://github.com/dart-lang/sdk/issues/46218?

**Comment by lrhn** [comment]
Such an annotation would be introduced by the analyzer and likely be in `package:meta`.

If we want to get the functionality for platform declarations, we can use a `@pragma("analyzer:await_not_required")` annotation.
(Heck, the declaration of `awaitNotRequired` could be just `const awaitNotRequired = pragma('analyzer:await_not_required');`, then everyone can use the same value as annotation.)
